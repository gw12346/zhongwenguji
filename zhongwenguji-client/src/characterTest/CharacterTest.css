.TestCardStack {
  display: flex;
  justify-content: center;
}

.TestCard {
  opacity: 1;
  transform: translateX(0) translateY(0) rotateZ(0);
  font-size: 16px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  border: 3px solid black;
  box-sizing: border-box;
  width: 12em;
  height: 15em;
  border-radius: 3px;
  overflow-y: auto;
  background-color: white;
  position: absolute;
}

.TestCard--showDefinition {
  justify-content: flex-start;
}

.TestCard__character {
  font-size: 8em;
  align-self: center;
}

.TestCard__definition {
  word-break: break-word;
  font-size: 1.5em;
  margin: 0 0.5em;
}

.TestCard__pinyin {
  font-size: 2em;
  min-height: 1.5em;
  line-height: 1.5em;
  margin: 0.2em 0;
}

.cardSwipe-leave {
  transform: translateX(0);
}

/* Leave animation */

.TestCard--markedUnknown.cardSwipe-leave,
.TestCard--markedKnown.cardSwipe-leave
{
  opacity: 0.01;
  transition: transform 300ms ease-in-out, opacity 300ms ease-in-out;
}

.TestCard--markedUnknown.cardSwipe-leave.cardSwipe-leave-active {
  background-color: #fee;
  transform: translateX(-150%) translateY(50%) rotateZ(-10deg);
}

.TestCard--markedKnown.cardSwipe-leave.cardSwipe-leave-active {
  background-color: #efe;
  transform: translateX(150%) translateY(50%) rotateZ(10deg);
}

/* Enter animation */
.TestCard--markedUnknown.cardSwipe-enter,
.TestCard--markedKnown.cardSwipe-enter {
  opacity: 0.001;
  z-index: 1;
}


.TestCard--markedUnknown.cardSwipe-enter {
  background-color: #fee;
  transform: translateX(-150%) translateY(50%) rotateZ(-10deg);
}

.TestCard--markedKnown.cardSwipe-enter {
  background-color: #efe;
  transform: translateX(150%) translateY(50%) rotateZ(10deg);
}

.TestCard--markedUnknown.cardSwipe-enter.cardSwipe-enter-active,
.TestCard--markedKnown.cardSwipe-enter.cardSwipe-enter-active
{
  transition: transform 100ms ease-in, opacity 100ms ease-in;
  opacity: 1;
  transform: translateX(0) translateY(0) rotateZ(0);
}