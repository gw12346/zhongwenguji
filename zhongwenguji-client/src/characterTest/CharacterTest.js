import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  toggleDefinition,
  markCurrentKnown,
  markCurrentUnknown,
  undoDiscard
} from './characterTestActions';
import {
  currentCard,
  isShowDefinition,
  status,
  scoreStatistics,
  resultData
} from './characterTestReducer';
import mapSelectors from '../util/mapSelectors';
import keyHandler from '../util/keyHandler';
import CardStackMobile from '../card/CardStackMobile';
// import CardStackDesktop from '../card/CardStackDesktop';
import style from './CharacterTest.module.scss';

const noop = () => {};

const GRAPH_PADDING = 10;

class CharacterTest extends Component {
  static propTypes = {
    currentCard: PropTypes.object,
    isShowDefinition: PropTypes.bool,
    status: PropTypes.string,
    scoreStatistics: PropTypes.object,
    resultData: PropTypes.object,

    toggleDefinition: PropTypes.func,
    markCurrentKnown: PropTypes.func,
    markCurrentUnknown: PropTypes.func,
    undoDiscard: PropTypes.func,
  };

  static defaultProps = {
    currentCard: null,
    isShowDefinition: false,
    status: '',
    scoreStatistics: {},
    resultData: {},

    toggleDefinition: noop,
    markCurrentKnown: noop,
    markCurrentUnknown: noop,
    undoDiscard: noop,
  };

  componentDidMount() {
    this.keyHandler = keyHandler({
      onKeyDown: this.handleKeyDown
    });
  }

  componentWillUnmount() {
    this.keyHandler.unregister();
  }

  handleKeyDown = (key) => {
    switch (key) {
      case 'ArrowDown':
        this.props.toggleDefinition();
        break;
      case 'ArrowLeft':
        this.props.markCurrentUnknown();
        break;
      case 'ArrowRight':
        this.props.markCurrentKnown();
        break;
      case 'ArrowUp':
        this.props.undoDiscard();
        break;
      default:
        break;
    }
  };

  handleDiscardLeft = () => {
    this.props.markCurrentUnknown();
  };

  handleDiscardRight = () => {
    this.props.markCurrentKnown();
  };

  handleUndo = () => {
    this.props.undoDiscard();
  };

  handleDefinition = () => {
    this.props.toggleDefinition();
  };

  renderLiveStats() {
    const padding = 10;
    const barWidth = 10;
    const { sectionStats = [] } = this.props.scoreStatistics;

    return (
      <svg width="420" height="220">
        {sectionStats.map(({ isTested, knownPercent, range }, i) => (<g key={i}>
          {(isTested && knownPercent >= 0) ? (
            <circle
              cx={padding + barWidth * i}
              cy={padding + 100 - knownPercent}
              r={3}
              style={{fill: 'rgb(0,0,255)'}}
            />
          ) : null}
        </g>))}
      </svg>
    );
  }

  renderResults() {
    const {
      curvePoints = [],
      samplePoints = [],
      knownEstimate = -1,
      knownEstimateUncertainty = -1
    } = this.props.resultData;

    return (
      <div>
        <div>You know {knownEstimate} 汉字 ± {knownEstimateUncertainty}</div>
        <div>
          <svg width={1000 + 2 * GRAPH_PADDING} height={100 + 2 * GRAPH_PADDING}>
            <g>
              {curvePoints.map(([x1, y1], i) => {
                const next = curvePoints[i + 1];
                if (next) {
                  const [x2, y2] = next;
                  return (
                    <line
                      key={x1}
                      x1={GRAPH_PADDING + x1 / 10}
                      y1={GRAPH_PADDING + 100 - y1}
                      x2={GRAPH_PADDING + x2 / 10}
                      y2={GRAPH_PADDING + 100 - y2}
                      style={{stroke: 'rgb(255,0,0)', 'strokeWidth': 1}}
                    />
                  );
                }
                return null;
              })}
            </g>
            <g>
              {samplePoints.map(([x, y]) => (
                <circle
                  key={x}
                  cx={GRAPH_PADDING + x / 10}
                  cy={GRAPH_PADDING + 100 - y}
                  r={3}
                  style={{fill: 'rgb(0,0,255)'}}
                />
              ))}
            </g>
          </svg>
        </div>
      </div>
    );
  }

  render() {
    return (
      <div className={style.container}>
        {this.props.resultData ? this.renderResults() : (
          <CardStackMobile
            currentCard={this.props.currentCard}
            showDefinition={this.props.isShowDefinition}
            onDiscardLeft={this.handleDiscardLeft}
            onDiscardRight={this.handleDiscardRight}
            onUndo={this.handleUndo}
            onDefinition={this.handleDefinition}
          />
        )}
      </div>
    );
  }
}

export { CharacterTest as Pure };

export default connect(mapSelectors({
  currentCard,
  isShowDefinition,
  status,
  scoreStatistics,
  resultData
}), {
  toggleDefinition,
  markCurrentKnown,
  markCurrentUnknown,
  undoDiscard
})(CharacterTest);

