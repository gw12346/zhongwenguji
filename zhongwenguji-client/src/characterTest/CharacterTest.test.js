import React from 'react';
import renderer from 'react-test-renderer';
import { Pure as CharacterTest } from './CharacterTest';

const mockUnregister = jest.fn();
let mockOnKeyDown = null;

jest.mock('../util/keyHandler', () => jest.fn(({ onKeyDown } = {}) => {
  mockOnKeyDown = onKeyDown;
  return { unregister: mockUnregister };
}));

describe('CharacterTest', () => {
  beforeEach(() => {
    mockUnregister.mockClear();
  });

  afterEach(() => {
    mockOnKeyDown = null;
  });

  let component;

  function setup(props) {
    component = renderer.create(
      <CharacterTest {...props} />
    );
  }

  it('renders', () => {
    setup();
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('toggles the definition', () => {
    const toggleDefinition = jest.fn();
    setup({ toggleDefinition });
    mockOnKeyDown('ArrowDown');
    expect(toggleDefinition).toHaveBeenCalled();
  });

  it('marks the card unknown', () => {
    const markCurrentUnknown = jest.fn();
    setup({ markCurrentUnknown });
    mockOnKeyDown('ArrowLeft');
    expect(markCurrentUnknown).toHaveBeenCalled();
  });

  it('marks the card known', () => {
    const markCurrentKnown = jest.fn();
    setup({ markCurrentKnown });
    mockOnKeyDown('ArrowRight');
    expect(markCurrentKnown).toHaveBeenCalled();
  });

  it('un-does the previous marking', () => {
    const undoDiscard = jest.fn();
    setup({ undoDiscard });
    mockOnKeyDown('ArrowUp');
    expect(undoDiscard).toHaveBeenCalled();
  });
});
