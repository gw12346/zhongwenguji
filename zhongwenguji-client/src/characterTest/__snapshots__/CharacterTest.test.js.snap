// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterTest renders 1`] = `
<div
  className="container"
>
  <div>
    <div>
      You know 
      -1
       汉字 ± 
      -1
    </div>
    <div>
      <svg
        height={120}
        width={1020}
      >
        <g />
        <g />
      </svg>
    </div>
  </div>
</div>
`;
